### PR TITLE
infoschema: add metric database/table to query cluster metric table.

### PR DESCRIPTION
**Pull request by crazycs520**
### What problem does this PR solve? 
* Add add metric database/table to query cluster metric table.
* Add 2 session variable to control the query metric time range:
    * `tidb_metric_schema_range_duration`, the query duration, it will affect the generated promQL.
    * `tidb_metric_schema_step`, the query step time, it will affect the step value when query promQL.

release test PR:https://github.com/pingcap/tidb-test/pull/968, merge test first.

```sql
mysql>use METRIC_SCHEMA;
mysql>select * from up;
+----------------------------+-------+-----------------+-------------------+
| time                       | value | instance        | job               |
+----------------------------+-------+-----------------+-------------------+
| 2019-12-04 18:46:33.398000 | 1.0   | 127.0.0.1:10080 | tidb              |
| 2019-12-04 18:46:33.398000 | 0.0   | 127.0.0.1:10081 | tidb              |
| 2019-12-04 18:46:33.398000 | 0.0   | 127.0.0.1:10082 | tidb              |
| 2019-12-04 18:46:33.398000 | 0.0   | 127.0.0.1:10083 | tidb              |
| 2019-12-04 18:46:33.398000 | 0.0   | 127.0.0.1:10090 | tidb              |
| 2019-12-04 18:46:33.398000 | 0.0   | 127.0.0.1:10091 | tidb              |
| 2019-12-04 18:46:33.398000 | 0.0   | 127.0.0.1:10092 | tidb              |
| 2019-12-04 18:46:33.398000 | 0.0   | 127.0.0.1:10093 | tidb              |
| 2019-12-04 18:46:33.398000 | 1.0   | 127.0.0.1:20180 | tikv              |
| 2019-12-04 18:46:33.398000 | 0.0   | 127.0.0.1:20181 | tikv              |
| 2019-12-04 18:46:33.398000 | 0.0   | 127.0.0.1:20182 | tikv              |
| 2019-12-04 18:46:33.398000 | 0.0   | 127.0.0.1:20183 | tikv              |
| 2019-12-04 18:46:33.398000 | 0.0   | 127.0.0.1:20184 | tikv              |
| 2019-12-04 18:46:33.398000 | 1.0   | 127.0.0.1:2379  | pd                |
| 2019-12-04 18:46:33.398000 | 0.0   | 127.0.0.1:2479  | pd                |
| 2019-12-04 18:46:33.398000 | 0.0   | 127.0.0.1:2579  | pd                |
| 2019-12-04 18:46:33.398000 | 0.0   | 127.0.0.1:9100  | overwritten-nodes |
| 2019-12-04 18:46:33.398000 | 1.0   | localhost:9090  | prometheus        |
+----------------------------+-------+-----------------+-------------------+
mysql>select * from query_duration;
+----------------------------+------------------+----------+----------+----------+
| time                       | value            | instance | sql_type | quantile |
+----------------------------+------------------+----------+----------+----------+
| 2019-12-04 18:46:37.134000 | 0.00430970543826 |          |          | 0.9      |
+----------------------------+------------------+----------+----------+----------+
```

You can use `explain` to check the generated promQL too:

```sql
mysql>desc select * from query_duration;
+----------------+----------+------+------------------------------------------------------------------------------------------------------------+
| id             | count    | task | operator info                                                                                              |
+----------------+----------+------+------------------------------------------------------------------------------------------------------------+
| MemTableScan_4 | 10000.00 | root | promQL:histogram_quantile(0.9, sum(rate(tidb_server_handle_query_duration_seconds_bucket{}[60s])) by (le)) |
+----------------+----------+------+------------------------------------------------------------------------------------------------------------+
```

### What is changed and how it works?

Actually, we use `promQL` to query metric data. Every metric table has a related `promQL`. 
Such as `query_duration` metric table, there is a definition in code:

```go
	"query_duration": {
		promQL:   `histogram_quantile($QUANTILE, sum(rate(tidb_server_handle_query_duration_seconds_bucket{$LABEL_CONDITION}[$RANGE_DURATION])) by (le))`,
		labels:   []string{"instance", "sql_type"},
		quantile: 0.90,
	},
	"up": {
		promQL: `up{$LABEL_CONDITION}`,
		labels: []string{"instance", "job"},
	},
```  

Then TiDB will generate a metric table as below:

```sql
mysql>desc query_duration;
+----------+-------------------+------+-----+-------------------+-------+
| Field    | Type              | Null | Key | Default           | Extra |
+----------+-------------------+------+-----+-------------------+-------+
| time     | datetime unsigned | YES  |     | CURRENT_TIMESTAMP |       |
| value    | double unsigned   | YES  |     | <null>            |       |
| instance | varchar(512)      | YES  |     | <null>            |       |
| sql_type | varchar(512)      | YES  |     | <null>            |       |
| quantile | double unsigned   | YES  |     | 0.9               |       |
+----------+-------------------+------+-----+-------------------+-------+
mysql>desc up;
+----------+-------------------+------+-----+-------------------+-------+
| Field    | Type              | Null | Key | Default           | Extra |
+----------+-------------------+------+-----+-------------------+-------+
| time     | datetime unsigned | YES  |     | CURRENT_TIMESTAMP |       |
| value    | double unsigned   | YES  |     | <null>            |       |
| instance | varchar(512)      | YES  |     | <null>            |       |
| job      | varchar(512)      | YES  |     | <null>            |       |
+----------+-------------------+------+-----+-------------------+-------+

``` 

As you can see, there is 3 variable in `promQL`, it will be replaced when execute SQL query:

```go
$QUANTILE  
$LABEL_CONDITION
$RANGE_DURATION
```

If you execute below SQL:

```
select * from query_duration where time > "2019-11-25 00:00:00" and time < "2019-11-25 00:01:00" and sql_type="general";
``` 

It will generated a promQL as below:

```go
histogram_quantile(0.90, sum(rate(tidb_server_handle_query_duration_seconds_bucket{sql_type="general"}[60s])) by (le))
```

The `$QUANTILE ` and `$RANGE_DURATION` is `0.90`, `60s`, it's the session variable value of `tidb_metric_schema_range_duration`.

Then send the `promQL`  to PD to query metric data which between in the "2019-11-25 00:00:00" and "2019-11-25 00:01:00".

**Attension**

Actually the variable in the `promQL` will always use the default value, use the value specified in SQL will be finished in the next PR.

### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)

Code changes

Side effects

Related changes

Release note

 - Write release note for bug-fix or new feature.

### CODE REVIEW

File: domain/domain.go [hunk 1]
```diff
@@ -34,6 +34,7 @@ import (
 	"github.com/pingcap/tidb/ddl"
 	"github.com/pingcap/tidb/domain/infosync"
 	"github.com/pingcap/tidb/infoschema"
+	"github.com/pingcap/tidb/infoschema/metrictable"
```
Comment (lonng) [inline domain/domain.go]: How about metricschema

File: infoschema/infoschema_test.go [hunk 2]
```diff
@@ -122,13 +122,13 @@ func (*testSuite) TestT(c *C) {
 	is := handle.Get()
 
 	schemaNames := is.AllSchemaNames()
-	c.Assert(schemaNames, HasLen, 3)
-	c.Assert(testutil.CompareUnorderedStringSlice(schemaNames, []string{infoschema.Name, "PERFORMANCE_SCHEMA", "Test"}), IsTrue)
+	c.Assert(schemaNames, HasLen, 4)
+	c.Assert(testutil.CompareUnorderedStringSlice(schemaNames, []string{infoschema.Name, "METRIC", "PERFORMANCE_SCHEMA", "Test"}), IsTrue)
```
Comment (lonng) [inline infoschema/infoschema_test.go]: It's better to define a constant for `METRICS_SCHEMA`.

Comment (crazycs520) [reply infoschema/infoschema_test.go]: Great, done.

File: infoschema/metricschema/table.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
+package metricschema
+
+import (
+	"github.com/pingcap/tidb/sessionctx"
```
Comment (djshow832) [inline infoschema/metricschema/table.go]: Please format this import.

File: sessionctx/variable/varsutil.go [hunk 4]
```diff
@@ -655,6 +655,15 @@ func ValidateSetSystemVar(vars *SessionVars, name string, value string) (string,
 			}
 		}
 		return formatVal, nil
+	case TiDBMetricSchemaStep, TiDBMetricSchemaRangeDuration:
+		v, err := strconv.ParseInt(value, 10, 64)
+		if err != nil {
+			return value, ErrWrongValueForVar.GenWithStackByArgs(name)
+		}
+		if v < 10 || v > 60*60*60 {
```
Comment (djshow832) [inline sessionctx/variable/varsutil.go]: You can use `checkUInt64SystemVar(name, value, 10, 60*60*60, vars)`

Comment (crazycs520) [reply sessionctx/variable/varsutil.go]: `checkUInt64SystemVar` will not return error when the `v` is not valid.

File: meta/autoid/autoid.go [hunk 5]
```diff
@@ -41,6 +41,8 @@ const (
 	InformationSchemaDBID int64 = SystemSchemaIDFlag | 1
 	// PerformanceSchemaDBID is the performance_schema schema id, it's exports for test.
 	PerformanceSchemaDBID int64 = SystemSchemaIDFlag | 10000
+	// MetricSchemaDBID is the metric_schema schema id, it's exports for test.
```
Comment (djshow832) [inline meta/autoid/autoid.go]:
```suggestion
	// MetricSchemaDBID is the metric_schema schema id, it's exported for test.
```

File: sessionctx/variable/sysvar.go [hunk 6]
```diff
@@ -727,6 +727,8 @@ var defaultSysVars = []*SysVar{
 	{ScopeGlobal, TiDBEvolvePlanTaskEndTime, DefTiDBEvolvePlanTaskEndTime},
 	{ScopeGlobal | ScopeSession, TiDBIsolationReadEngines, "tikv,tiflash,tidb"},
 	{ScopeGlobal | ScopeSession, TiDBStoreLimit, strconv.FormatInt(atomic.LoadInt64(&config.GetGlobalConfig().TiKVClient.StoreLimit), 10)},
+	{ScopeSession, TiDBMetricSchemaStep, strconv.Itoa(DefTiDBMetricSchemaStep)},
```
Comment (djshow832) [inline sessionctx/variable/sysvar.go]: Why is `ScopeGlobal` not supported?

Comment (crazycs520) [reply sessionctx/variable/sysvar.go]: I think there is no scene that needs the global scope variable.

File: infoschema/metricschema/init.go [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+	initOnce := func() {
+		dbID := autoid.MetricSchemaDBID
+		tableID := dbID + 1
+		metricTables := make([]*model.TableInfo, 0)
+		for name, def := range metricTableMap {
+			cols := def.genColumnInfos()
+			tableInfo := buildTableMeta(name, cols)
+			tableInfo.ID = tableID
+			tableID++
+			for i, c := range tableInfo.Columns {
```
[the remark this answers is not in the document]
Comment (crazycs520) [reply infoschema/metricschema/init.go]: Great.

File: infoschema/metricschema/query.go [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/store/tikv"
+	"github.com/prometheus/client_golang/api"
+	"github.com/prometheus/client_golang/api/prometheus/v1"
+	pmodel "github.com/prometheus/common/model"
+)
+
+const promReadTimeout = time.Second * 10
```
[the remark this answers is not in the document]
Comment (crazycs520) [reply infoschema/metricschema/query.go]: em... Currently, dial timeout for `pd`,`tikv`, and `read timeout` for `tikv` are both not configurable, so I think it's ok to use a const value here.

**Comment by crazycs520** [comment]
@djshow832 , Great, Thanks, Addressed, PTAL

File: infoschema/metricschema/table.go [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+				record := def.genRecord(m.Metric, v, r)
+				rows = append(rows, record)
+			}
+		}
+	}
+	return rows
+}
+
+func (def *metricTableDef) genRecord(metric pmodel.Metric, pair pmodel.SamplePair, r promQLQueryRange) []types.Datum {
+	record := make([]types.Datum, 0, 8)
```
[the remark this answers is not in the document]
Comment (crazycs520) [reply infoschema/metricschema/table.go]: Already changed.

File: sessionctx/variable/varsutil.go [hunk 10]
```diff
@@ -660,6 +660,15 @@ func ValidateSetSystemVar(vars *SessionVars, name string, value string) (string,
 			}
 		}
 		return formatVal, nil
+	case TiDBMetricSchemaStep, TiDBMetricSchemaRangeDuration:
+		v, err := strconv.ParseInt(value, 10, 64)
+		if err != nil {
+			return value, ErrWrongValueForVar.GenWithStackByArgs(name)
```
Comment (lamxTyler) [inline sessionctx/variable/varsutil.go]: This error needs 2 args, the `value` is also needed.

File: infoschema/metricschema/tables.go [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+	meta *model.TableInfo
+	cols []*table.Column
+}
+
+func tableFromMeta(alloc autoid.Allocator, meta *model.TableInfo) (table.Table, error) {
+	return createmetricSchemaTable(meta), nil
+}
+
+// createmetricSchemaTable creates all metricSchemaTables
+func createmetricSchemaTable(meta *model.TableInfo) *metricSchemaTable {
```
Comment (lamxTyler) [inline infoschema/metricschema/tables.go]: `createMetrics...`

Comment (crazycs520) [reply infoschema/metricschema/tables.go]: done.

File: sessionctx/variable/varsutil.go (hunk 10 above)
Comment (crazycs520) [reply sessionctx/variable/varsutil.go]: great, done.

File: executor/builder.go [hunk 12]
```diff
@@ -1242,24 +1243,36 @@ func (b *executorBuilder) getStartTS() (uint64, error) {
...
-	default:
-		tb, _ := b.is.TableByID(v.Table.ID)
-		e = &TableScanExec{
-			baseExecutor:   newBaseExecutor(b.ctx, v.Schema(), v.ExplainID()),
-			t:              tb,
-			columns:        v.Columns,
-			seekHandle:     math.MinInt64,
-			isVirtualTable: !tb.Type().IsNormalTable(),
+		switch v.Table.Name.L {
+		case strings.ToLower(infoschema.TableClusterConfig):
```
Comment (djshow832) [inline executor/builder.go]: It's not good, table name `cluster_config` can also exist in other system databases in theory. 
I suggest remove `if util.IsMemOrSysDB(v.DBName.L)`,  and check both schema name and table name for each table.

Comment (crazycs520) [reply executor/builder.go]: great, done.

File: util/misc.go [hunk 13]
```diff
@@ -141,12 +141,16 @@ const (
 	PerformanceSchemaName = "PERFORMANCE_SCHEMA"
 	// PerformanceSchemaLowerName is the `PERFORMANCE_SCHEMA` database lower name.
 	PerformanceSchemaLowerName = "performance_schema"
+	// MetricSchemaName is the `METRIC_SCHEMA` database name.
+	MetricSchemaName = "METRIC_SCHEMA"
```
Comment (lonng) [inline util/misc.go]: Why not use a global variable which type is `model.CIStr` .

Comment (crazycs520) [reply util/misc.go]: Go does not support struct constants.

File: infoschema/metricschema/table.go [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+// TODO: read from system table.
+var metricTableMap = map[string]MetricTableDef{
+	"query_duration": {
+		PromQL:   `histogram_quantile($QUANTILE, sum(rate(tidb_server_handle_query_duration_seconds_bucket{$LABEL_CONDITION}[$RANGE_DURATION])) by (le))`,
+		Labels:   []string{"instance", "sql_type"},
+		Quantile: 0.90,
+	},
+	"up": {
```
Comment (lonng) [inline infoschema/metricschema/table.go]:
```suggestion
	"uptime": {
```

File: util/misc.go (hunk 13 above)
Comment (lonng) [reply util/misc.go]: How about use `var`?

File: executor/builder.go [hunk 15]
```diff
@@ -1259,24 +1260,35 @@ func (b *executorBuilder) getStartTS() (uint64, error) {
...
+			e = &ClusterReaderExec{
+				baseExecutor: newBaseExecutor(b.ctx, v.Schema(), v.ExplainID()),
+				retriever: &clusterConfigRetriever{
+					extractor: v.Extractor.(*plannercore.ClusterConfigTableExtractor),
+				},
+			}
 		}
 	}
+	if e != nil {
```
Comment (lonng) [inline executor/builder.go]: This check is weak (if someday we use a builder to construct the executor and builder will return nil, which will break this assumption), prefers to use the `default` case.

File: executor/metric_reader.go [hunk 16]
```diff
@@ -0,0 +1,198 @@
...
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/parser/model"
+	"github.com/pingcap/parser/mysql"
+	"github.com/pingcap/tidb/infoschema/metricschema"
+	"github.com/pingcap/tidb/store/tikv"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/prometheus/client_golang/api"
+	v1 "github.com/prometheus/client_golang/api/prometheus/v1"
```
Comment (lonng) [inline executor/metric_reader.go]: `v1` is not a good alias.

File: executor/metric_reader.go [hunk 17]
```diff
@@ -0,0 +1,198 @@
...
+	mutableRow := chunk.MutRowFromTypes(retTypes(e))
+	for _, row := range rows {
+		mutableRow.SetDatums(row...)
+		req.AppendRow(mutableRow.ToRow())
+	}
+	return nil
+}
+
+func (e *MetricReaderExec) getRows() (fullRows [][]types.Datum, err error) {
+	tblDef, err := metricschema.GetMetricTableDef(e.table.Name.L)
```
Comment (lonng) [inline executor/metric_reader.go]: Why get the table definition in `getRows`? I think we should get the definition in the executor build stage.

File: executor/metric_reader.go [hunk 18]
```diff
@@ -0,0 +1,198 @@
...
+		return nil, err
+	}
+
+	queryClient, err := newQueryClient(addr)
+	if err != nil {
+		return nil, err
+	}
+
+	promQLAPI := v1.NewAPI(queryClient)
+	ctx, cancel := context.WithTimeout(context.Background(), promReadTimeout)
```
Comment (lonng) [inline executor/metric_reader.go]: It's better to use the `ctx` received from `Next` method.

File: infoschema/metricschema/table.go (hunk 14 above)
Comment (crazycs520) [reply infoschema/metricschema/table.go]: The metric value is not a time, just a bool value( 0 or 1), indicate the instance whether up in the time.

File: executor/metric_reader.go (hunk 17 above)
Comment (crazycs520) [reply executor/metric_reader.go]: Ye, but the executor building stage can't return error, so I move this in here.

File: executor/metric_reader.go (hunk 18 above)
Comment (crazycs520) [reply executor/metric_reader.go]: great, done.

File: util/misc.go (hunk 13 above)
Comment (crazycs520) [reply util/misc.go]: done.

File: executor/builder.go (hunk 15 above)
Comment (crazycs520) [reply executor/builder.go]: done.

File: executor/metric_reader.go (hunk 17 above)
Comment (lonng) [reply executor/metric_reader.go]: How about in `Open`?

File: executor/metric_reader.go [hunk 19]
```diff
@@ -0,0 +1,198 @@
...
+	}
+	rows := make([][]types.Datum, len(fullRows))
+	for i, fullRow := range fullRows {
+		row := make([]types.Datum, len(e.outputCols))
+		for j, col := range e.outputCols {
+			row[j] = fullRow[col.Offset]
+		}
+		rows[i] = row
+	}
+	return rows, nil
```
Comment (lonng) [inline executor/metric_reader.go]: Why not doing this in `getRows` directly?

File: executor/metric_reader.go [hunk 20]
```diff
@@ -0,0 +1,198 @@
...
+
+func (e *MetricReaderExec) getMetricAddr() (string, error) {
+	// Get PD servers info.
+	store := e.ctx.GetStore()
+	etcd, ok := store.(tikv.EtcdBackend)
+	if !ok {
+		return "", errors.Errorf("%T not an etcd backend", store)
+	}
+	for _, addr := range etcd.EtcdAddrs() {
+		addr = strings.TrimSpace(addr)
```
Comment (lonng) [inline executor/metric_reader.go]: Is necessary to trim space?

File: executor/metric_reader.go [hunk 21]
```diff
@@ -0,0 +1,198 @@
...
+	}
+	return &queryClient{
+		promClient,
+	}, nil
+}
+
+// URL implement the api.Client interface.
+// This is use to convert prometheus api path to PD API path.
+func (c *queryClient) URL(ep string, args map[string]string) *url.URL {
+	ep = strings.Replace(ep, "api/v1", "pd/api/v1/metric", 1)
```
Comment (lonng) [inline executor/metric_reader.go]: Do we have a chance to set the request header `req.Header.Add("PD-Allow-follower-handle", "true")`?

File: infoschema/infoschema.go [hunk 22]
```diff
@@ -102,9 +102,8 @@ type InfoSchema interface {
 }
 
 // Information Schema Name.
-const (
-	Name      = util.InformationSchemaName
-	LowerName = util.InformationSchemaLowerName
+var (
+	Name = util.InformationSchemaName.O
```
Comment (lonng) [inline infoschema/infoschema.go]: Can we remove this global variable and replace it with `util.InformationSchemaName.O`?

File: infoschema/metricschema/init.go [hunk 23]
```diff
@@ -0,0 +1,95 @@
...
+		for name, def := range metricTableMap {
+			cols := def.genColumnInfos()
+			tableInfo := buildTableMeta(name, cols)
+			tableInfo.ID = tableID
+			tableID++
+			metricTables = append(metricTables, tableInfo)
+		}
+		dbInfo := &model.DBInfo{
+			ID:      dbID,
+			Name:    model.NewCIStr(util.MetricSchemaName.O),
```
Comment (lonng) [inline infoschema/metricschema/init.go]:
```suggestion
			Name:    util.MetricSchemaName.O,
```

File: infoschema/perfschema/const.go [hunk 24]
```diff
@@ -16,9 +16,8 @@ package perfschema
 import "github.com/pingcap/tidb/util"
 
 // Performance Schema Name.
-const (
-	Name      = util.PerformanceSchemaName
-	LowerName = util.PerformanceSchemaLowerName
+var (
+	Name = util.PerformanceSchemaName.O
```
Comment (lonng) [inline infoschema/perfschema/const.go]: ditto

File: server/http_handler_test.go [hunk 25]
```diff
@@ -688,7 +688,7 @@ func (ts *HTTPHandlerTestSuite) TestGetSchema(c *C) {
 	var dbs []*model.DBInfo
 	err = decoder.Decode(&dbs)
 	c.Assert(err, IsNil)
-	expects := []string{"information_schema", "mysql", "performance_schema", "test", "tidb"}
+	expects := []string{"information_schema", "metric_schema", "mysql", "performance_schema", "test", "tidb"}
```
Comment (lonng) [inline server/http_handler_test.go]: I am not sure whether `metric_schema` is the proper name, or is `metrics_schema` better? @bb7133

File: util/misc.go [hunk 26]
```diff
@@ -132,21 +133,19 @@ func SyntaxWarn(err error) error {
...
-	// InformationSchemaLowerName is the `INFORMATION_SCHEMA` database lower name.
-	InformationSchemaLowerName = "information_schema"
+	InformationSchemaName = model.CIStr{O: "INFORMATION_SCHEMA", L: "information_schema"}
 	// PerformanceSchemaName is the `PERFORMANCE_SCHEMA` database name.
-	PerformanceSchemaName = "PERFORMANCE_SCHEMA"
-	// PerformanceSchemaLowerName is the `PERFORMANCE_SCHEMA` database lower name.
-	PerformanceSchemaLowerName = "performance_schema"
+	PerformanceSchemaName = model.CIStr{O: "PERFORMANCE_SCHEMA", L: "performance_schema"}
+	// MetricSchemaName is the `METRIC_SCHEMA` database name.
+	MetricSchemaName = model.CIStr{O: "METRIC_SCHEMA", L: "metric_schema"}
```
Comment (lonng) [inline util/misc.go]: Prefer using `model.NewCIStr("...")` to initialize the variable.

Comment (crazycs520) [reply util/misc.go]: I prefer too, but can't do this.....

File: executor/metric_reader.go (hunk 17 above)
Comment (crazycs520) [reply executor/metric_reader.go]: done.

File: executor/metric_reader.go (hunk 19 above)
Comment (crazycs520) [reply executor/metric_reader.go]: ?

File: executor/metric_reader.go (hunk 21 above)
Comment (crazycs520) [reply executor/metric_reader.go]: I'm afraid not ...

File: infoschema/metricschema/init.go (hunk 23 above)
Comment (crazycs520) [reply infoschema/metricschema/init.go]: great, done.

File: executor/metric_reader.go [hunk 27]
```diff
@@ -0,0 +1,202 @@
...
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/prometheus/client_golang/api"
+	promv1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	pmodel "github.com/prometheus/common/model"
+)
+
+const promReadTimeout = time.Second * 10
+
+// MetricReaderExec uses to read metric data.
+type MetricReaderExec struct {
```
Comment (lonng) [inline executor/metric_reader.go]: I think there is no need to add an individual executor for this function, we can reuse `ClusterReaderExec`  and implement a retriever is enough.

File: sessionctx/variable/tidb_vars.go [hunk 28]
```diff
@@ -359,6 +359,12 @@ const (
 
 	// TiDBStoreLimit indicates the limit of sending request to a store, 0 means without limit.
 	TiDBStoreLimit = "tidb_store_limit"
+
+	// TiDBMetricSchemaStep indicates the step when query metric schema.
+	TiDBMetricSchemaStep = "tidb_metric_schema_step"
+
+	// TiDBMetricSchemaRangeDuration indicates the range duration when query metric schema.
+	TiDBMetricSchemaRangeDuration = "tidb_metric_schema_range_duration"
```
Comment (lonng) [inline sessionctx/variable/tidb_vars.go]: How about `tidb_metrics_query_range_duration`

File: infoschema/metricschema/table.go [hunk 29]
```diff
@@ -0,0 +1,115 @@
...
+	"strings"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/parser/mysql"
+	"github.com/pingcap/tidb/sessionctx"
+)
+
+const (
+	promQLQuantileKey       = "$QUANTILE"
+	promQLLabelConditionKey = "$LABEL_CONDITION"
```
Comment (lonng) [inline infoschema/metricschema/table.go]:
```suggestion
	promQLLabelConditionKey = "$LABEL_CONDITIONS"
```

File: executor/metric_reader.go (hunk 21 above)
Comment (crazycs520) [reply executor/metric_reader.go]: done.

File: executor/metric_reader.go (hunk 27 above)
Comment (crazycs520) [reply executor/metric_reader.go]: great, done.

File: sessionctx/variable/tidb_vars.go [hunk 30]
```diff
@@ -359,6 +359,12 @@ const (
 
 	// TiDBStoreLimit indicates the limit of sending request to a store, 0 means without limit.
 	TiDBStoreLimit = "tidb_store_limit"
+
+	// TiDBMetricSchemaStep indicates the step when query metric schema.
+	TiDBMetricSchemaStep = "tidb_metric_schema_step"
```
[the remark this answers is not in the document]
Comment (crazycs520) [reply sessionctx/variable/tidb_vars.go]: great, done

File: executor/cluster_reader.go [hunk 31]
```diff
@@ -36,7 +36,7 @@ import (
 )
 
 type clusterRetriever interface {
-	retrieve(ctx sessionctx.Context) ([][]types.Datum, error)
+	retrieve(sctx sessionctx.Context, ctx context.Context) ([][]types.Datum, error)
```
Comment (lonng) [inline executor/cluster_reader.go]: Please place the `context.Context` before `sessionctx.Context`.

Comment (crazycs520) [reply executor/cluster_reader.go]: great, done.

**Pull request merged**